### ISSUE TITLE
Support Windows Builds

### ISSUE THREAD

**Issue by DavidDeSimone** [emacs-ng/emacs-ng#47]
Currently, I strongly suspect that windows builds are broken, though I have not verified. I know that there is code is `ng_async.rs` that will not compile with windows due to references to `std::fs::unix::{...}`. We will have to compile that functionality out of ng_async, but that will not effect major features like Deno or WebRender.

**Comment by DavidDeSimone** [comment]
Since the days of remacs, this has become both easier and harder.

The main challenge is linking v8, which comes as a precompiled .lib, made with the msvc tool chain. While mingw gcc can handling linking .lib files, I've found no way to have it fulfill the linking requirements for its C++ dependencies. A basic application yields link errors for __CxxFrameHandler3 and the managed name for typeinfo . A more complex library like v8 yields a massive amount of errors.

I think the correct way to do this is to just make deno a dll, and compile with the rust windows gnu toolchain. I haven't tested that yet, but it's where my head is at.

**Comment by DavidDeSimone** [comment]
Update: I'm personally pausing on working on this, but I will summarize my findings:

Rust has two primary Windows toolschains, the msvc toolchain, and the pc-gnu toolchain. Building emacs-ng without JavaScript is straight forward, how the primary issue is that v8 is precompiled as a .lib file under the MSVC toolchain. However, compiling emacs-ng under  the MSVC toolchain causes issues, because including rust libraries compiled under MSVC with GCC seems to cause missing symbol issues I'm unable to resolve.

I have successfully tested that we can solve this by splitting v8 at at the DLL barrier. If we make v8 a DLL compiled under MSVC, we can include it in a program compiled via gcc and linked via ld. You will just need to include the v8 DLL in the src/ directory that emacs-ng is built in. The hard part is how to not make our build system a nightmare over this:

Making the v8 DLL is easy - make a dummy crate with a dylib export called "v8_dll", include rusty_v8 and export it. You will produce a DLL with the needed code. Now you can use that code within a pc-gnu compiled application. 

The hard part is that we need to be under the msvc toolchain while making the DLL, and then SWITCH to the pc-gnu toolchain to build libemacsng.a and ultimately emacs-ng itself. We can perform this with toolchain directory files, but it will add complexity to an already complex (and slow) build process. 

I'm open to any suggestions if I've missed anything critical. If anyone else wants to grab this in the meantime, feel free. Otherwise I'm going to put this on pause until we hit other critical project milestones, or there is stronger user demand.

**Comment by kiennq** [comment]
@DavidDeSimone Thanks for your work to make this happens on Windows.
Can you give some pointer (step by step would be great) on how to build `emacs-ng` on Windows?
I would like to try this and not really familiar with gnu toolchain on Windows.

**Comment by DavidDeSimone** [comment]
@kiennq I'm going to do some work today to either 1. Make windows builds work and upload a prebuilt package or 2. Better document what I’ve done so far

**Comment by DavidDeSimone** [comment]
Update: I still don't have windows working. What you need to do to be able to set yourself up:

NOTE: Following these steps will not get you to a working build. This is just to document progress. 
1. Download msys2
2. Follow the instructions in nt/INSTALL , which lists deps that you need to download pacman
3. Download rust. 
4. Download clang, install under default install path
5. Checkout https://github.com/DavidDeSimone/emacs-webrender/tree/windows

The main issue is that we need to compile that I get linker errors when compiling our rust code under the msvc toolchain (missing calls like CxxFrameHandler3, etc), and I cannot compile deno under the pc-windows-gnu toolchain (rusty_v8 only seems to support the msvc toolchain).

My ideas for way ahead now:

1. Statically link all the required cxx/c msvc libraries. This might work, but I think it will result in a violation of the GPL, so I don't think it would be long term viable.
2. Compile v8 under the gnu toolchain, and then compile and link deno under the gnu toolchain.
3. Refactor our code so that we can cleanly split deno into a DLL compiled under msvc, while compiling any code that references emacs functions into a library that can be compiled under pc-gnu. This means that we need to call deno functions via extern "C", which is less then ideal. 

I will add a big disclaimer here that I am not a Windows expert, so if anything in this thread looks wrong, or anyone reading this has something to contribute, please add in to the discussion.

**Comment by DavidDeSimone** [comment]
Splitting deno at the dll barrier didn't have the effect that I thought it would. I think the fundamental mistake that I'm making is that v8 needs certain msft libraries (that include things like ___CxxStackFrame3, etc). that we need to include somehow. The magic is that cargo/rustc handles this behind the scenes via the build environment and dynamic linking. My current conclusion is that I need to figure out what cargo is doing and implement some to all of that in our build system when it comes to setting up dynamic linking to the dlls that v8 needs. I also need to not violate the GPLv3 in that process.

**Comment by yyoncho** [comment]
> I also need to not violate the GPLv3 in that process.

can you elaborate?

**Comment by DavidDeSimone** [comment]
> > I also need to not violate the GPLv3 in that process.
> 
> 
> 
> can you elaborate?

One of the recommendations I got in response to the errors I was seeing was to statically link certain Microsoft provided libs, however my understanding is that linking those libraries would be a violation of the GPL. However I believe there are alternative solutions that are GPL compatible

**Comment by kiennq** [comment]
>  was to statically link certain Microsoft provided libs, however my understanding is that linking those libraries would be a violation of the GPL
I think Microsoft only provides stubs libs, which is simply call forwarding and load necessary dll if needed. Does that still count as violating GPL?

**Comment by DavidDeSimone** [comment]
> >  was to statically link certain Microsoft provided libs, however my understanding is that linking those libraries would be a violation of the GPL
> 
> 
> 
> I think Microsoft only provides stubs libs, which is simply call forwarding and load necessary dll if needed. Does that still count as violating GPL?

I don't believe what you described is a violation. What people were recommending to me was actually statically linking code with no dll loading, which would be a violation.

**Comment by DavidDeSimone** [comment]
I am going to do some experimenting with https://github.com/denoland/rusty_v8/pull/690 - I think that may help resolve our link issues.

**Comment by cannero** [comment]
Hello @DavidDeSimone, is there anything I can help here with?
One thing I've tried is to compile rusty_v8 with the buffer security check disabled, /GS-. This reduced the number of unresolved external symbols, but I'm not sure if it helps in any way. My idea was to remove all references to the msvc crt, but this seems not possible.

**Comment by DavidDeSimone** [comment]
Honestly I hit kind of roadblock as documented before. I know this is
possible, but I think it's just difficult to conceptualize for me due
to the number of build systems are are using simultaneously. One
though I had was to build temacs with main being defined in rust, and
using `cargo build` to build temacs. Doing this while linking rusty_v8
doesn't produce these linking errors in a toy example, but I didn't
make a full POC. Talking through this we could look into what cargo
build is doing - if they can make it work, so should we. I think we
just need a little more experimentation with both linker settings and
executable setup.
